### PR TITLE
Add batch tx simulation for zk overflow issue

### PR DESCRIPTION
**Pull request by amit-momin**
In certain product use cases, tx simulation is needed for multiple requests at the same time. To minimize the response time for simulation, a batch method was added to simulate multiple transactions at the same time.

### CODE REVIEW

File: core/chains/evm/client/chain_client.go [hunk 1]
```diff
@@ -272,10 +279,57 @@ func (c *chainClient) LatestFinalizedBlock(ctx context.Context) (*evmtypes.Head,
...
+	if err != nil {
+		return err
+	}
+
+	for _, elem := range rpcBatchCalls {
+		params := elem.Args[0].(map[string]interface{})
+		for i := 0; i < len(reqs); i++ {
+			req := &reqs[i]
+			// Match the request to rpc response to set the error in the proper object
+			if params["from"] == req.From &&
```
Comment (dimriou) [inline core/chains/evm/client/chain_client.go:326]: What is this check for? IIUC, you're checking req against params, which comes from rpcBatchCalls.Args and it should it always be true, as it doesn't get mutated, only rpcBatchCalls.Result does.

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go:326]: Originally I was using it as a sanity check to make sure I was setting the error in the right req but you're right. The order doesn't change so this nested looping doesn't serve a purpose. I'll get this changed.

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go]: Updated this in the following commit

File: core/chains/evm/client/chain_client.go [hunk 2]
```diff
@@ -272,10 +279,48 @@ func (c *chainClient) LatestFinalizedBlock(ctx context.Context) (*evmtypes.Head,
...
+	reqs := []TxSimulationRequest{
+		{
+			From: from,
+			To:   &to,
+			Data: data,
+		},
+	}
+	err := c.BatchCheckTxValidity(ctx, reqs)
+	if err != nil {
+		return NewSendError(err)
```
Comment (dimriou) [inline core/chains/evm/client/chain_client.go:291]: `BatchCheckTxValidity` can fail for multiple reasons, for example the multi node might not be able to select a proper active node. That fail will not be related to the error classification. Perhaps it makes sense to return two values, SendError & error so the callers can tell if the call succeeded or not.

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go:291]: I was considering that originally but was reluctant to change the method signature product teams are currently using. The main reason being product teams aren't doing more than just a nil check on the error which they can do to the SendError as well. Guess that's an unnecessary limitation on the error handling though so I can split this into 2 error.

Comment (amit-momin) [reply core/chains/evm/client/chain_client.go]: Pushed the change in the following commit

File: core/chains/evm/client/chain_client_test.go [hunk 3]
```diff
@@ -69,3 +69,250 @@ func TestChainClient_BatchCallContext(t *testing.T) {
...
+		require.NoError(t, err)
+
+		// No error for first request
+		require.Empty(t, reqs[0].Error)
+
+		// No Out-of-counter error for second request
+		require.Equal(t, errorMsg, reqs[1].Error.Error())
+	})
+
+	t.Run("returns the proper errors to the associated requests", func(t *testing.T) {
```
Comment (dimriou) [inline core/chains/evm/client/chain_client_test.go:255]: I think this functionality is already tested in `TestChainClient_BatchCallContext`

Comment (amit-momin) [reply core/chains/evm/client/chain_client_test.go]: True. This is left over from when I had the request/batchElem matching which is removed now. I've deleted this in the latest commit.

**Review by dimriou — APPROVED** [review]

**Pull request closed without merging**